metadata-ingestion/src/datahub/ingestion/source/schema_inference/object.py

```diff
@@ -0,0 +1,181 @@
+from collections import Counter
+from typing import Any
+from typing import Counter as CounterType
+from typing import Dict, Iterable, Tuple, Union
+
+from mypy_extensions import TypedDict
+
+
+class BasicSchemaDescription(TypedDict):
+    types: CounterType[type]  # field types and times seen
+    count: int  # times the field was seen
+
+
+class SchemaDescription(BasicSchemaDescription):
+    delimited_name: str  # collapsed field name
+    # we use 'mixed' to denote mixed types, so we need a str here
+    type: Union[type, str]  # collapsed type
+    nullable: bool  # if field is ever missing
+
+
+def is_nullable_doc(doc: Dict[str, Any], field_path: Tuple) -> bool:
+    """
+    Check if a nested field is nullable in a document from a collection.
+
+    Parameters
+    ----------
+        doc:
+            document to check nullability for
+        field_path:
+            path to nested field to check, ex. ('first_field', 'nested_child', '2nd_nested_child')
+    """
+
+    field = field_path[0]
```

Comment:
Is it safe to access the first element? Defensive
```python
if not field_path:
   return True
```
```

Reply:
We should be guaranteed that the field_path exists by construction, but it can't hurt to have the check